src/include/migraphx/op/reduce_op.hpp

```diff
@@ -94,16 +94,75 @@ struct reduce_op : op_name<Derived>
         return {{"normalize_axes", normalize}, {"reduce", true}};
     }
 
-    std::vector<int64_t> tune_axes(std::size_t n_dim) const
+    shape collapse_reduced_axes(const shape& original_shape,
+                                const std::vector<int64_t>& reduce_axes) const
     {
-        auto tuned_axes = axes;
-        if(tuned_axes.empty())
+        auto lens = original_shape.lens();
+        for(const auto a : reduce_axes)
         {
-            tuned_axes.resize(n_dim);
-            std::iota(tuned_axes.begin(), tuned_axes.end(), 0);
+            lens[a] = 1;
         }
 
-        return tuned_axes;
+        return original_shape.with_lens(lens);
+    }
+
+    // Compute the output shape for cases when the input tensor has a dynamic shape.
+    //
+    // If the axes are passed as an attribute, or a constant input, we can precisely
+    // determine which axes need to be collapsed even during parsing.
+    //
+    // If the axes are passed as a variable input(indicated by an empty axes attribute), we cannot
+    // determine which axes must be collapsed until we see the actual input values, so we must treat
+    // each axis as potentially collapsable and set its minimum dimension to 1.
+    shape compute_dynamic_shape(const std::vector<shape>& inputs) const
+    {
+        const auto& data_shape = inputs[0];
+        auto dims              = data_shape.dyn_dims();
+        if(axes.empty())
+        {
+            for(auto& dim : dims)
+            {
+                dim = {1, dim.max};
+            }
+        }
+        else
+        {
+            for(auto a : axes)
+            {
+                dims[a] = {1, 1};
+            }
+        }
+
+        return {data_shape.type(), dims};
+    }
+
+    // Compute the output shape for cases when the input tensor has a static shape.
+    // Depending on how axes is passed to the operator the output shape can be either dynamic or
+    // static.
+    //
+    // If the axes are passed as an attribute, or a constant input, we can precisely
```

Comment:
We cant produce a static shape with a constant input because the input to compute_shape is just the shapes.